src/source.cpp

```diff
@@ -367,9 +375,18 @@ void load_custom_source_library()
   // reset errors
   dlerror();
 
-  // get the function from the library
-  using sample_t = Particle::Bank (*)(uint64_t* seed);
-  custom_source_function = reinterpret_cast<sample_t>(dlsym(custom_source_library, "sample_source"));
+  if (custom_source_parameters.empty()) {
+    // get the function from the library
+    using sample_t = Particle::Bank (*)(uint64_t* seed);
+    custom_source_function = reinterpret_cast<sample_t>(dlsym(custom_source_library, "sample_source"));
+  } else {
+    // get the functions to create and destroy the CustomSource from the library
+    create_custom_source_t* create_custom_source = (create_custom_source_t*) dlsym(custom_source_library, "create");
+    destroy_custom_source = (destroy_custom_source_t*) dlsym(custom_source_library, "destroy");
+
+    // create a pointer to an instance of the CustomSource
+    custom_source = create_custom_source(custom_source_parameters.c_str());
+  }
```

Comment:
I mostly went with the two routes to avoid breaking any backward compatibility for users that might be actively using the existing `sample_source` function import. Agree that the single approach is cleaner... if @makeclean and @Shimwell are ok with removing the `sample_source` approach (I suspect they will be :)) and they're the only ones using that then I'm happy to clean that up.

Reply:
One way is fine, just having the user write a super simple class which inherits from CustomSource like 
```
#include "openmc/source.h"

extern "C"  source* create() {
  return std::unique_ptr<sample>();
}

class sample : custom source

particle::bank sample::sample_source() {
 do stuff
}
```

I think is simple enough I think go for it, and rename the serialised_source -> source since noone will know what serialisation is.